src/eins-network-id.c

```diff
@@ -0,0 +1,238 @@
+                          (guint8) (gateway_ip >> 16),
+                          (guint8) (gateway_ip >> 24));
+}
+
+#define PROC_IPV4_ARP_FILE "/proc/net/arp"
+#define IPV4_ARP_REGEXP "^%s\\s+(\\S+\\s+){2}(?P<HW>[0-9a-f:]{17})"
+
+/*
+ * find the given IPv4 address in the ARP cache to find the HW address.
+ * relies on two assumptions:
+ *  - the first (obviously) is that the IP has been communicated with recently
+ *    enough to be in the cache. for an internet-connected host, this is a
+ *    pretty safe bet for the gateway IP.
+ *  - the second is that if the host has multiple interfaces which are in the
+ *    same IPv4 segment, that these are the same physical network, because the
+ *    search is not constrained by interface which would be the strictly
+ *    correct thing to do. this configuration seems pretty unlikely (two bonded
+ *    load-balanced links with separate gateways?) on a client machine, and for
+ *    our purposes it doesn't matter if we always report one gateway, or report
+ *    both over time, so it's not worth getting excited about.
+ */
+gchar *
+get_ipv4_hwaddr (const gchar *address)
+{
+  g_autofree gchar *arp = NULL;
+
+  if (!g_file_get_contents (PROC_IPV4_ARP_FILE, &arp, NULL, NULL) || !arp)
+    {
+      g_debug ("could not load " PROC_IPV4_ARP_FILE);
+      return NULL;
+    }
+
+  g_autofree gchar *arp_rx_raw = g_strdup_printf (IPV4_ARP_REGEXP, address);
+  g_autoptr(GRegex) arp_rx = g_regex_new (arp_rx_raw, G_REGEX_MULTILINE, 0, NULL);
+  g_autoptr(GMatchInfo) arp_mi = NULL;
+
+  if (!g_regex_match (arp_rx, arp, 0, &arp_mi) || !arp_mi)
+    {
+      g_debug ("couldn't find HW address in " PROC_IPV4_ARP_FILE);
+      return NULL;
+    }
+
+  const gchar *hw = g_match_info_fetch_named (arp_mi, "HW");
+  g_assert (hw);
+  return g_ascii_strdown (hw, -1);
+}
+
+#define PROC_IPV6_ROUTE_FILE "/proc/net/ipv6_route"
+#define IPV6_ROUTE_REGEXP "^(0{32}\\s00\\s){2}(?P<Gateway>[0-9a-f]{32})"
+
+/*
+ * greps the /proc/net/ipv6_route file for the first IPv6 default route. same
+ * assumptions as above apply.
+ */
+gchar *
+get_ipv6_default_gateway (void)
+{
+  g_autofree gchar *route = NULL;
+
+  if (!g_file_get_contents (PROC_IPV6_ROUTE_FILE, &route, NULL, NULL) || !route)
+    {
+      g_debug ("could not load " PROC_IPV6_ROUTE_FILE);
+      return NULL;
+    }
+
+  g_autoptr(GRegex) route_rx = g_regex_new (IPV6_ROUTE_REGEXP, G_REGEX_MULTILINE, 0, NULL);
+  g_autoptr(GMatchInfo) route_mi = NULL;
+
+  if (!g_regex_match (route_rx, route, 0, &route_mi) || !route_mi)
+    {
+      g_debug ("couldn't match IPv6 default gateway in " PROC_IPV6_ROUTE_FILE);
+      return NULL;
+    }
+
+  const gchar *gateway = g_match_info_fetch_named (route_mi, "Gateway");
+  g_assert (gateway);
+
+  guint32 addr[8];
+  if (sscanf(gateway,"%4x%4x%4x%4x%4x%4x%4x%4x",
+             &addr[0], &addr[1], &addr[2], &addr[3],
+             &addr[4], &addr[5], &addr[6], &addr[7]) < 6)
+    {
+      g_assert_not_reached ();
```

Comment:
This could do with a comment saying that the regexp has already validated the format of `gateway`, so this should always succeed.